include/swift/ABI/Task.h

```diff
@@ -148,19 +150,93 @@ class AsyncTask : public HeapObject, public Job {
   /// Reserved for the use of the task-local stack allocator.
   void *AllocatorPrivate[4];
 
+  enum class TaskLocalChainStatus : uintptr_t {
+      /// This task is known to be a "terminal" node in the lookup of task locals.
+      /// In other words, even if it had a parent, the parent (and its parents)
+      /// are known to not contain any any more task locals, and thus any further
+      /// search beyond this task.
+      IsEmptyTerminal   = 0b00,
+      /// The storage pointer points at the next TaskLocalChainItem in this task.
+      IsNextItemPointer = 0b01,
+      /// The storage pointer points at a parent AsyncTask,
+      /// in which we should continue the lookup.
+      IsParentPointer   = 0b11
+  };
+
+  struct TaskLocalChainItem {
+      /// Mask used for the low status bits in a task local chain item.
+      static const uintptr_t statusMask = 0x03;
+
+      /// Metadata* to the Key of the task local value
+      uintptr_t keyType;
+      /// OpaqueValue* to the Value of the task local value
+      HeapObject *storage;
+
+      // TaskLocalChainItem*
+//      uintptr_t next;
+  };
+
+  /// Reserved for a chain of task local values,
+  /// if empty, the search should continue in this task's parent (if it has one).
+  TaskLocalChainItem *TaskLocal;
+
   AsyncTask(const HeapMetadata *metadata, JobFlags flags,
             TaskContinuationFunction *run,
             AsyncContext *initialContext)
     : HeapObject(metadata), Job(flags, run),
       ResumeContext(initialContext),
-      Status(ActiveTaskStatus()) {
+      Status(ActiveTaskStatus()),
+      TaskLocal(nullptr) {
     assert(flags.isAsyncTask());
   }
 
   void run(ExecutorRef currentExecutor) {
     ResumeTask(this, currentExecutor, ResumeContext);
   }
 
+  HeapObject* localGet(const Metadata *keyType) {
+    assert(keyType && "Task.Local key must not be null.");
+
+    if (!TaskLocal) {
+      return nullptr;
+    }
+
+    fprintf(stderr, "error: %s [%s:%d] TaskLocal=%d keyType=%d, lookup keyType=%d\n",
+            __FUNCTION__, __FILE_NAME__, __LINE__,
+            TaskLocal, TaskLocal->keyType, reinterpret_cast<uintptr_t>(keyType));
+    // TODO: can the contained list/dictionary for the key
+    if (TaskLocal->keyType == reinterpret_cast<uintptr_t>(keyType)) {
```

Comment:
[3] during the read however... we see that the TaskLocal has the right values, however...

It seems that the TaskLocal->storage now points at garbage; so I assume the object got deallocated... however no amount of swift_retaining it seems to matter; what am I missing here?